core/src/com/unciv/ui/utils/UnitGroup.kt

```diff
@@ -4,30 +4,103 @@ import com.badlogic.gdx.graphics.Color
 import com.unciv.ui.images.ImageGetter
 import com.unciv.ui.utils.extensions.center
+import com.unciv.ui.utils.extensions.centerX
 import com.unciv.ui.utils.extensions.surroundWithCircle
+import com.unciv.ui.utils.extensions.toGroup
 
 class UnitGroup(val unit: MapUnit, val size: Float): Group() {
-    var blackSpinningCircle: Image? = null
     var actionGroup :Group? = null
     val unitBaseImage = ImageGetter.getUnitIcon(unit.name, unit.civInfo.nation.getInnerColor())
-        .apply { setSize(size * 0.75f, size * 0.75f) }
-    var background: Image? = null
+        .apply {
+            if (unit.isCivilian())
+                setSize(size * 0.60f, size * 0.60f)
+            else
+                setSize(size * 0.75f, size * 0.75f) }
+    var flagSelection: Image? = null
+    var flagHalo: Image? = null
+    var flagBg: Group? = null
 
     init {
-        background = getBackgroundImageForUnit()
-        background?.apply {
-            this.color = unit.civInfo.nation.getOuterColor()
-            this.color.a = UncivGame.Current.settings.unitIconOpacity
-            setSize(size, size)
+
+        val bgW = size
+        var bgH = size
+
+        if (unit.isFortified())
+            bgH *= 1.1f
+        else if (unit.isEmbarked())
+            bgH *= 1.1f
+        else if (unit.isCivilian())
+            bgH *= 1.2f
+
+        setSize(bgW*1.25f, bgH*1.25f)
+
+        flagSelection = getBackgroundSelectionForUnit()
+        flagSelection?.apply {
+            align = Align.center
+            color.a = 0f
+            if (unit.isCivilian() && !unit.isEmbarked())
+                setSize(bgW*2.5f, bgH*2.5f)
+            else
+                setSize(bgW*2f, bgH*2f)
+            center(this@UnitGroup)
+        }
+
+        flagHalo = getBackgroundImageForUnit()
+        flagHalo?.apply{
+            align = Align.center
+            color.a = 0f
+            if (unit.isCivilian() && !unit.isEmbarked())
+                setSize(bgW*1.40f, bgH*1.40f)
+            else
+                setSize(bgW*1.30f, bgH*1.30f)
+            center(this@UnitGroup)
+        }
+
+        flagBg = Group().apply {
+            setSize(bgW*1.15f, bgH*1.15f)
+            val outer = getBackgroundImageForUnit().apply {
+                align = Align.center
+                color = unit.civInfo.nation.getInnerColor()
+                color.a = UncivGame.Current.settings.unitIconOpacity
+                if (unit.isCivilian() && !unit.isEmbarked())
+                    setSize(bgW*1.20f, bgH*1.20f)
+                else
+                    setSize(bgW*1.15f, bgH*1.15f)
+            }
+
+            val inner = getBackgroundImageForUnit().apply {
+                align = Align.center
+                color = unit.civInfo.nation.getOuterColor()
+                color.a = UncivGame.Current.settings.unitIconOpacity
+                setSize(bgW, bgH)
+            }
+
+            val mask = getBackgroundMaskForUnit()?.apply {
+                align = Align.center
```

Comment:
We have *double* apply's here, which is even worse. Please flatten to like
 - define group
 - Define group parts
 - Add parts to group

Reply:
Done